File: sdks/flutter/test/webhooks_test.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:test/test.dart';
import 'package:neuratalk_sdk/neuratalk_sdk.dart';

void main() {
  group('verifyWebhookSignature', () {
    test('accepts a valid signature', () {
      const secret = 'topsecret';
      final body = utf8.encode('{"event":"call.completed"}');
      final sig = Hmac(sha256, utf8.encode(secret)).convert(body).toString();
      expect(verifyWebhookSignature(body, sig, secret), isTrue);
    });

    test('accepts a valid signature with sha256= prefix', () {
      const secret = 'topsecret';
      final body = utf8.encode('{"event":"call.completed"}');
      final sig = Hmac(sha256, utf8.encode(secret)).convert(body).toString();
      expect(verifyWebhookSignature(body, 'sha256=$sig', secret), isTrue);
    });

    test('rejects an invalid signature', () {
      final body = utf8.encode('body');
      expect(verifyWebhookSignature(body, 'deadbeef' * 8, 'secret'), isFalse);
    });

    test('rejects a tampered body', () {
      const secret = 'topsecret';
      final sig = Hmac(sha256, utf8.encode(secret)).convert(utf8.encode('original')).toString();
      expect(verifyWebhookSignature(utf8.encode('tampered'), sig, secret), isFalse);
    });
  });
}

File: sdks/kotlin/src/test/kotlin/com/neuratalk/sdk/WebhooksTest.kt
package com.neuratalk.sdk

import java.nio.charset.StandardCharsets
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class WebhooksTest {
    private fun hmacHex(secret: String, body: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
        return mac.doFinal(body.toByteArray(StandardCharsets.UTF_8)).joinToString("") { "%02x".format(it) }
    }

    @Test
    fun `accepts a valid signature`() {
        val secret = "topsecret"
        val body = """{"event":"call.completed"}"""
        val sig = hmacHex(secret, body)
        assertTrue(verifyWebhookSignature(body, sig, secret))
    }

    @Test
    fun `accepts a valid signature with sha256 prefix`() {
        val secret = "topsecret"
        val body = """{"event":"call.completed"}"""
        val sig = hmacHex(secret, body)
        assertTrue(verifyWebhookSignature(body, "sha256=$sig", secret))
    }

    @Test
    fun `rejects an invalid signature`() {
        assertFalse(verifyWebhookSignature("body", "deadbeef".repeat(8), "secret"))
    }

    @Test
    fun `rejects a tampered body`() {
        val secret = "topsecret"
        val sig = hmacHex(secret, "original")
        assertFalse(verifyWebhookSignature("tampered", sig, secret))
    }
}
